File: questions/q3/solution.py
import sys

# ============================================
# IMPLEMENT YOUR SOLUTION IN THE main() FUNCTION
# ============================================

def main(guide, moves):
    """
    Calculate your total score in Rock-Paper-Scissors based on a strategy guide.
    
    Game Rules:
    - R (Rock) beats S (Scissors)
    - S (Scissors) beats P (Paper)
    - P (Paper) beats R (Rock)
    
    Scoring:
    - Win: 6 points
    - Draw: 3 points
    - Loss: 1 point
    
    Args:
        guide: List of 6 strings representing 3 pairs [opponent_move, your_response]
               e.g., ["R", "P", "P", "S", "S", "R"]
        moves: List of strings representing opponent's move sequence
               e.g., ["R", "P", "S"]
    
    Your solution should print the result to stdout using print()
    """
    # ============================================
    # YOUR CODE HERE
    # ============================================
    
    m = {}
    for i in range(0, len(guide) - 1, 2):
        m[guide[i]] = guide[i + 1]

    score = 0
    for move in moves:
        if move == "R":
            if m[move] == "P":
                score += 1
            elif m[move] == "S":
                score -= 1
        elif move == "P":
            if m[move] == "S":
                score += 1
            elif m[move] == "R":
                score -= 1
        elif move == "S":
            if m[move] == "R":
                score += 1
            elif m[move] == "P":
                score -= 1

    print(score)
    
    # ============================================
    # END YOUR CODE
    # ============================================


# ============================================
# DO NOT MODIFY BELOW THIS LINE
# ============================================
if __name__ == "__main__":
    data = sys.stdin.read().strip()
    words = data.split()
    guide = words[:6]
    moves = words[6:]
    main(guide, moves)

File: questions/q3/Solution.java
package q3;

import java.util.*;

// ============================================
// IMPLEMENT YOUR SOLUTION IN THE solve() METHOD
// ============================================

public class Solution {
    /**
     * Calculate your total score in Rock-Paper-Scissors based on a strategy guide.
     * 
     * Game Rules:
     * - R (Rock) beats S (Scissors)
     * - S (Scissors) beats P (Paper)
     * - P (Paper) beats R (Rock)
     * 
     * Scoring:
     * - Win: 6 points
     * - Draw: 3 points
     * - Loss: 1 point
     * 
     * @param guide Array of 6 strings representing 3 pairs [opponent_move, your_response]
     * @param moves Array of strings representing the opponent's move sequence
     * 
     * Your solution should print the result to stdout using System.out.println()
     */
    public static void solve(String[] guide, String[] moves) {
        // ============================================
        // YOUR CODE HERE
        // ============================================
        
        // Parse guide
        Map<String, String> guideMap = new HashMap<>();
        for (int i = 0; i < guide.length; i += 2) {
            guideMap.put(guide[i], guide[i + 1]);
        }
        
        // Calculate score
        int score = 0;
        for (String move : moves) {
            String response = guideMap.get(move);
            
            if (move.equals("R")) {
                if (response.equals("P")) score += 1;
                else if (response.equals("S")) score -= 1;
            } else if (move.equals("P")) {
                if (response.equals("S")) score += 1;
                else if (response.equals("R")) score -= 1;
            } else if (move.equals("S")) {
                if (response.equals("R")) score += 1;
                else if (response.equals("P")) score -= 1;
            }
        }
        
        System.out.println(score);
        
        // ============================================
        // END YOUR CODE
        // ============================================
    }
    
    // ============================================
    // DO NOT MODIFY BELOW THIS LINE
    // ============================================
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String line = scanner.nextLine();
        String[] words = line.trim().split("\\s+");
        
        String[] guide = Arrays.copyOfRange(words, 0, 6);
        String[] moves = Arrays.copyOfRange(words, 6, words.length);
        
        solve(guide, moves);
        scanner.close();
    }
}